proposals/csharp-7.0/tuples.md

```diff
@@ -0,0 +1,437 @@
+var t = (sum: 0, count: 1);
+t.sum   = 1;        // sum   is the name for the field #1
+t.Item1 = 1;        // Item1 is the name of underlying field #1 and is also available
+
+var t1 = (0, 1); // tuple omits the field names.
+t.Item1 = 1;     // underlying field name is still available
+
+t.ToString();     // ToString on the underlying tuple type is called.
+
+System.ValueTuple<int, int> vt = t; // identity conversion
+(int moo, int boo) t2 = vt;         // identity conversion
+
+```
+
+Because of the dual nature of tuples, it is not allowed to assign field names that overlap with preexisting member names of the underlying type. The only exception is the use of predefined `Item1`, `Item2`,...`ItemN` at corresponding position N, since that would not be ambiguous.
+
+```csharp
+var t =  (ToString: 0, ObjectEquals: 1);  // error: names match underlying member names
+var t1 = (Item1: 0, Item2: 1);            // valid
+var t2 = (misc: 0, Item1: 1);             // error: "Item1" was used in a wrong position
+```
+
+If the tuple is bigger than the limit of 7, the implementation will nest the "tail" as a tuple into the eighth element recursively. This nesting is visible by accessing the `Rest` field of a tuple, but that field my be hidden from e.g. auto-completion, just as the `ItemX` field names may be hidden but allowed when a tuple has named elements.
+
+A well formed "big tuple" will have names `Item1` etc. all the way up to the number of tuple elements, even though the underlying type doesn't physically have those fields directly defined. The same goes for the tuple returned from the `Rest` field, only with the numbers "shifted" appropriately.
+
+For tuple element names occurring in partial type declarations, the names must be the same.
+
+```csharp
+partial class C : IEnumerable<(string name, int age)> { ... }
+partial class C : IEnumerable<(string fullname, int)> { ... } // error: names must be specified and the same
+```
+
+When tuple elements names are used in overridden signatures, or implementations of interface methods, tuple element names in parameter and return types must be preserved. It is an error for the same generic interface to be inherited or implemented twice with identity convertible type arguments that have conflicting tuple element names
+
+> note:
+>
+> For the purpose of overloading, overriding and hiding, tuples of the same types and lengths as well as their underlying ValueTuple types are considered equivalent. All other differences are immaterial. When overriding a member it is permitted to use tuple types with same or different field names than in the base member.
+>
+> A situation where same field names are used for non-matching fields between base and derived member signatures, a warning is reported by the compiler.  
+
+```csharp
+class Base
+{
+    virtual void M1(ValueTuple<int, int> arg){...}
+}
+class Derived : Base
+{
+    override void M1((int c, int d) arg){...} // valid override, signatures are equivalent
+}
```

Comment:
Incorrect.  See [https://sharplab.io/#v2:CYLg1APgAgTAjAWAFDNgAgEIEMDOBTZAb2TVLSgGY0A3ASwCcAXAVywBtyAWNAWTgAoAau2Z4AKswAObPAB5aAO0YAaNIsYA+NFnoBzAJSEAvshNJ0AETz1a1PMDQhMuAkmJIy5KgHs79G8B4XLwC/OpoAMaq4cD62nqGRmgA9Mk07LQOvtYBeKo4tLoKWCz0eDjxQXgAjsy27HhKpqgwaFY2dsCtTu229mhEJGSUaNn+mUFQ3Hz8YUqRcNHzEXE6BsYpaew43ulsmaoA7joKirqjCmiS3jgFAEYyaAC2tDjM+KMAZmjFT0EA5BF/gMkGYgA](https://sharplab.io/#v2:CYLg1APgAgTAjAWAFDNgAgEIEMDOBTZAb2TVLSgGY0A3ASwCcAXAVywBtyAWNAWTgAoAau2Z4AKswAObPAB5aAO0YAaNIsYA+NFnoBzAJSEAvshNJ0AETz1a1PMDQhMuAkmJIy5KgHs79G8B4XLwC/OpoAMaq4cD62nqGRmgA9Mk07LQOvtYBeKo4tLoKWCz0eDjxQXgAjsy27HhKpqgwaFY2dsCtTu229mhEJGSUaNn+mUFQ3Hz8YUqRcNHzEXE6BsYpaew43ulsmaoA7joKirqjCmiS3jgFAEYyaAC2tDjM+KMAZmjFT0EA5BF/gMkGYgA)

Reply:
added to #3154